### PR TITLE
1088. Confusing Number II

### PR DESCRIPTION
**Pull request by altayhunoglu**
Resolves: https://github.com/spiralgo/algorithms/issues/228

As we understand here: https://github.com/spiralgo/algorithms/issues/156, if a number is generated using the numbers: 0, 1, 6, 8, 9 and :

 - if it does `not equal` itself when it is rotated 180 degrees, it is a `confusing number`. 
 
 - If it `equals` itself when it is rotated 180 degrees, it is a `strobogrammatic number`. 
 
 Therefore:

>  (Total numbers) - (strobogrammatic numbers) = (confusing numbers)

It means, we can use the code here https://github.com/spiralgo/algorithms/pull/323, and just subtract its output from the total count of all numbers generated using 0, 1, 6, 8, 9.

### CODE REVIEW

File: src/main/java/algorithms/curated170/hard/ConfusingNumber2.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        {'0', '0'},
+        {'6', '9'},
+        {'1', '1'},
+        {'9', '6'},
+        {'8', '8'}
+    };
+
+    int highLen;
+    int lowLen;
+    int currLen;
```
Comment (altayhunoglu) [inline src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:17]:
```suggestion

```

**Comment by ErdemT09** [comment]
image
I don't understand how it's so efficient despite its use of strings and such.

File: src/main/java/algorithms/curated170/hard/ConfusingNumber2.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    public int confusingNumberII(int n) {
+        String num = Integer.toString(n);
+        int res = findTotal(num);
+        StrobogrammaticNumber3 strobogrammaticNumber = new StrobogrammaticNumber3();
+        return res - strobogrammaticNumber.strobogrammaticInRange("1", num) - 1;
+    }
+    private int findTotal(String s) {
+        if (s.length() == 0) return 1;
+        char first = s.charAt(0);
+        int res = count(first) * (int) (Math.pow(5, s.length() - 1));
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:24]: What is being done here?

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:24]: findTotal is to calculate the total count of the numbers which can be generated using those five numbers: 0, 1, 6, 8, 9.
It is why we use 5 to calculate the permutation.

As the "s" is a limit for us, we have to calculate numbers less than "s".
In order to do that: count(first) returns the numbers less than the parameter (first) from this set: 0, 1, 6, 8, 9.

findTotal is recursive as we should do the calculation for each digit.

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:24]: When considering a number like 9999, we ignore the permutations that start with 9. But, when we calculate the next digits, we are actually calculating the number of permutations that start with 9. Pretty clear now.

File: src/main/java/algorithms/curated170/hard/ConfusingNumber2.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        if (first == '0' || first == '1' || first == '6' || first == '8' || first == '9') {
+            res += findTotal(s.substring(1));
+        }
+        return res;
+    }
+    private int count(Character c) {
+        int res = 0;
+        for (char[] p : reflectives) {
+            if (p[0] < c) res += 1;
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:33]:
```suggestion
            if (p[0] < c) res ++;
```

File: src/main/java/algorithms/curated170/hard/ConfusingNumber2.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        return res - strobogrammaticNumber.strobogrammaticInRange("1", num) - 1;
+    }
+    private int findTotal(String s) {
+        if (s.length() == 0) return 1;
+        char first = s.charAt(0);
+        int res = count(first) * (int) (Math.pow(5, s.length() - 1));
+        if (first == '0' || first == '1' || first == '6' || first == '8' || first == '9') {
+            res += findTotal(s.substring(1));
+        }
+        return res;
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/hard/ConfusingNumber2.java:28]: Writing it a bit more explicity.
```suggestion
        int reflectivePermutations = count(first) * (int) (Math.pow(5, s.length() - 1));
        if (first == '0' || first == '1' || first == '6' || first == '8' || first == '9') {
            reflectivePermutations += findTotal(s.substring(1));
        }
        return reflectivePermutations;
```

**Review by ErdemT09 — APPROVED** [review]
It's a good method of problem solving to reduce a problem to another one. I like how that's being done here. 
Approved.

**Comment by altayhunoglu** [comment]
> It's a good method of problem solving to reduce a problem to another one. I like how that's being done here.
> Approved.

Vielen Dank für die Codeüberprüfung und für deine wertvolle Zusammenarbeit.

Ich rufe bewusst eine Methode in einer anderen Klasse auf, um ein Beispiel für unsere Idee des "Lego-Bausteine" zu geben.

**Pull request merged**